Support collection signing verification

eqrx opened the pull request:
It was requested that builder supports the opportunistic signature
verification feature of ansible-galaxy. this commit adds one ~~two~~ new
flags. ~~`disable-gpg-verify` lets galaxy always ignore signatures
and is simply added verbatim to the ansible-galaxy invocation in
the container file.~~ `keyring` is the path of a GPG keyring file
that shall contain public keys that are used to verify signatures
of collections. This is forwarded to the ansible-galaxy command
by copying the given path into the build directory and passing
the path to the copied file from within the container to the
ansible-galaxy invocation

Commit:
Support keyring and disable-gpg-verify flags of ansible-galaxy

It was requested that builder supports the opportunistic signature
verification feature of ansible-galaxy. this commit adds two new
flags. `disable-gpg-verify` lets galaxy always ignore signatures
and is simply added verbatim to the ansible-galaxy invocation in
the container file. `keyring` is the path of a GPG keyring file
that shall contain public keys that are used to verify signatures
of collections. This is forwarded to the ansible-galaxy command
by copying the given path into the build directory and passing
the path to the copied file from within the container to the
ansible-galaxy invocation

Shrews commented on ansible_builder/cli.py:
```diff
@@ -80,6 +80,19 @@ def add_container_options(parser):
     )
 
+    # Keyring option copied over from ansible-galaxy. Gets passed to the invocations of it verbatim.
+    create_command_parser.add_argument(
```
Both of these new options should be present for both `create` and `build` subcommands. There is a `for` loop down below that adds some common options.

Shrews commented on ansible_builder/main.py:
```diff
@@ -39,7 +39,13 @@ def __init__(self,
                  no_cache=False,
-                 verbosity=constants.default_verbosity):
+                 verbosity=constants.default_verbosity,
+                 disable_gpg_verify=False,
```
This is a bit odd that the default does not match the CLI option default of `True`. Same in `Containerfile` init.

eqrx replied:
Actually a good point, thank you

Shrews commented on ansible_builder/steps.py:
```diff
@@ -43,14 +43,19 @@ def __init__(self):
+
+        install_opts = f"-r {requirements_naming} --collections-path {constants.base_collections_path}"
+        if keyring:
+            install_opts += " --keyring ./keyring.gog"
```
`.gog` != `.gpg`  :)

eqrx replied:
Oh no, https://www.gog.com/ if branded into my brain, sorry xD

Shrews reviewed (changes requested):
Some changes noted. Since I'm seeing about adding tests, I'll make the changes that I suggested.

Shrews commented on ansible_builder/cli.py:
```diff
@@ -80,6 +80,19 @@ def add_container_options(parser):
+    create_command_parser.add_argument(
+        '--disable-gpg-verify',
+        dest='disable_gpg_verify',
+        action='store_true',
```
We want the default to be to NOT verify collection signatures. This forces the user to use this option in order to prevent that.

If we want the user to specifically opt-in to verification, I think we'd have to go with an option like `--enable-collection-verification` (bikeshed the name) which defaults to `False`. If that option is not given, we use `--disable-gpg-verify` during collection installation. If it is given, then we avoid using that option and also pass the keyring value along. We also probably want to verify that a keyring value is actually given if we enable signature verification.

(edited for some wrong assumptions and confusing wording)

Shrews replied:
Resolved this by turning on sig verify if `--keyring` is given. No need for the other option.

Shrews committed:
Simplify to only --keyring

Shrews commented on ansible_builder/steps.py:
```diff
@@ -43,14 +43,19 @@ def __init__(self):
+        if keyring:
+            install_opts += " --keyring ./keyring.gpg"
+        else:
+            install_opts += " --disable-gpg-verify"
```
Looks like we'll have to use the [`ANSIBLE_GALAXY_DISABLE_GPG_VERIFY`](https://docs.ansible.com/ansible-core/devel/reference_appendices/config.html#galaxy-disable-gpg-verify) env var to control this since this option isn't going to be supported on older versions of Ansible.

Shrews committed:
Use env var to disable sig verification

Shrews committed:
Add some unit tests

Shrews committed:
Some integration tests and docs

Shrews commented:
There might still be a discussion around whether or not keyring should be specified in the EE definition file vs. a command line option. I personally don't see a reason why it couldn't just be a EE file value.

Shrews commented:
FYI, I'm uncertain if the relative path to the copied keyring (./keyring.gpg) is correct. Looking into it. 

Edit: tests modified to test that the keyring is copied to the user data dir (`_build`) in the build context dir.

Shrews committed:
Test keyring is copied to expected location

Shrews commented:
FYI, builder is affected by https://github.com/ansible/ansible/issues/77349 since it does `role install` before `collection install`.

eqrx reviewed (commented):
Reviewing PR created by me since @Shrews did a lot of work on it :) -> lgtm

Shrews commented:
> Reviewing PR created by me since @Shrews did a lot of work on it :) -> lgtm

FYI, this isn't exactly done yet. PR #https://github.com/ansible/ansible/pull/77026 was recently merged which adds a couple of new signature verification options (`--required-valid-signature-count` and `--ignore-signature-status-code`) which we will need to handle.

Commit:
Add cli flags required-valid-signature-count and ignore-signature-status-code

These were introduced by https://github.com/ansible/ansible/pull/77026
and allow customizing the signing verification behaviour of
ansible-galaxy.

Shrews commented on test/integration/test_create.py:
```diff
@@ -50,3 +52,81 @@ def test_create_streams_output_with_invalid_verbosity(cli, build_dir_and_ee_yml)
+    reqyml = tmpdir / "requirements.yml"
+    reqyml.write_text("\n".join(req))
+    cli(f'ansible-builder create -c {tmpdir} -f {eeyml} --output-filename Containerfile '
+        '--galaxy-ignore-signature-status-code 500 --galaxy-required-valid-signature-count 3')
```
Are these options valid without specifying a keyring? If not, we probably should have a check for that early on in builder that errors out quickly for the user.

Shrews commented on test/unit/test_steps.py:
```diff
@@ -16,3 +17,25 @@ def test_additional_build_steps(verb):
+
+
+def test_galaxy_install_steps_with_keyring():
+    steps = list(GalaxyInstallSteps("requirements.txt", constants.default_keyring_name, None, None))
```
Would be nice to see a unit test where these two values are not `None`.

Shrews commented on ansible_builder/cli.py:
```diff
@@ -143,6 +143,18 @@ def add_container_options(parser):
+                       help='A gpg status code to ignore during signature verification when installing with '
+                       'ansible-galaxy(for example, NO_PUBKEY).')
+        p.add_argument('--galaxy-required-valid-signature-count',
+                       help='The number of signatures that must successfully verify collections and rolesfrom '
```
`roles` are not verified. There's no support for that.

I almost feel like it would be better if we didn't describe this too much and leave the galaxy documentation to fully explain. Your first sentence might be enough, plus "see docs for ansible-galaxy for further explanation" or similar.

Shrews commented on ansible_builder/main.py:
```diff
@@ -371,7 +401,11 @@ def prepare_build_context(self):
+            self.steps.extend(GalaxyInstallSteps(CONTEXT_FILES['galaxy'],
+                                                 self.copied_galaxy_keyring,
+                                                 self.galaxy_ignore_signature_status_code,
+                                                 self.galaxy_required_valid_signature_count))
```
Mentioned in a comment elsewhere, but feels like we should probably verify `galaxy-keyring` is given if one or both of these other options are supplied. I don't think they're valid without it.

Shrews commented on docs/usage.rst:
```diff
@@ -75,6 +75,21 @@ To use a definition file named something other than
 
    $ ansible-builder build --file=my-ee.yml
 
+``--galaxy-keyring``
```
Need some docs for the other new options.

eqrx commented:
retest

Shrews commented on ansible_builder/main.py:
```diff
@@ -294,7 +309,16 @@ class Containerfile:
+                 galaxy_ignore_signature_status_code=None):
+        """
+        :param str galaxy_keyring: GPG keyring file used by ansible-galaxy to opportunistically validate collection signatures.
+        :param str galaxy_required_valid_signature_count: Number of sigs (prepend + to disallow no sig( required for ansible-galaxy to accept collections.
```
```suggestion
        :param str galaxy_required_valid_signature_count: Number of sigs (prepend + to disallow no sig) required for ansible-galaxy to accept collections.
```

Shrews commented on ansible_builder/steps.py:
```diff
@@ -43,14 +43,36 @@ def __init__(self):
+        """Assumes given requirements file name and keyring has been placed in the build context.
+
+        :param str galaxy_keyring: GPG keyring file used by ansible-galaxy to opportunistically validate collection signatures.
+        :param str galaxy_required_valid_signature_count: Number of sigs (prepend + to disallow no sig( required for ansible-galaxy to accept collections.
```
```suggestion
        :param str galaxy_required_valid_signature_count: Number of sigs (prepend + to disallow no sig) required for ansible-galaxy to accept collections.
```

Shrews commented on docs/usage.rst:
```diff
@@ -75,6 +75,47 @@ To use a definition file named something other than
+``--galaxy-ignore-signature-status-code``
+*****************************************
+
+With --galaxy-keyring set it is possible to ignore certain errors happening while
```
```suggestion
With ``--galaxy-keyring`` set, it is possible to ignore certain errors that may occur while
```

Shrews commented on docs/usage.rst:
```diff
@@ -75,6 +75,47 @@ To use a definition file named something other than
+*****************************************
+
+With --galaxy-keyring set it is possible to ignore certain errors happening while
+verifying collections. It is passed unmodified to ansible-galaxy calls via the option
```
```suggestion
verifying collections. It is passed unmodified to ``ansible-galaxy`` calls via the option
```

Shrews commented on docs/usage.rst:
```diff
@@ -75,6 +75,47 @@ To use a definition file named something other than
+
+With --galaxy-keyring set it is possible to ignore certain errors happening while
+verifying collections. It is passed unmodified to ansible-galaxy calls via the option
+--ignore-signature-status-code. See the ansible-galaxy doc for more info.
```
```suggestion
``--ignore-signature-status-code``. See the ``ansible-galaxy`` documentation for more information.
```

Shrews commented on docs/usage.rst:
```diff
@@ -75,6 +75,47 @@ To use a definition file named something other than
+``--galaxy-required-valid-signature-count``
+*******************************************
+
+With --galaxy-keyring set it is possible specify how many valid signatures must present for each collection item.
```
```suggestion
With ``--galaxy-keyring`` set, it is possible to specify how many valid signatures must be present for each collection item.
```

Shrews commented on docs/usage.rst:
```diff
@@ -75,6 +75,47 @@ To use a definition file named something other than
+*******************************************
+
+With --galaxy-keyring set it is possible specify how many valid signatures must present for each collection item.
+Bear in mind that if this is set to an number collections having no signatures at all still pass.
```
```suggestion
```

I'd prefer to leave discussion of the details of what ansible-galaxy does with this value to the galaxy docs.

Shrews commented on docs/usage.rst:
```diff
@@ -75,6 +75,47 @@ To use a definition file named something other than
+
+With --galaxy-keyring set it is possible specify how many valid signatures must present for each collection item.
+Bear in mind that if this is set to an number collections having no signatures at all still pass.
+The value is passed unmodified to ansible-galaxy calls via the option
```
```suggestion
The value is passed unmodified to ``ansible-galaxy`` calls via the option
```

Shrews commented on docs/usage.rst:
```diff
@@ -75,6 +75,47 @@ To use a definition file named something other than
+With --galaxy-keyring set it is possible specify how many valid signatures must present for each collection item.
+Bear in mind that if this is set to an number collections having no signatures at all still pass.
+The value is passed unmodified to ansible-galaxy calls via the option
+--galaxy-required-valid-signature-count. See the ansible-galaxy doc for more info.
```
```suggestion
``--required-valid-signature-count``. See the ``ansible-galaxy`` documentation for more information.
```

Shrews commented on test/integration/test_create.py:
```diff
@@ -50,3 +52,83 @@ def test_create_streams_output_with_invalid_verbosity(cli, build_dir_and_ee_yml)
+    assert f"--keyring {constants.default_keyring_name}" in text
+
+
+def test_galaxy_signing_extra_args(cli, build_dir_and_ee_yml):
```
Let's add a similar test to `integration/test_build.py`.

Shrews reviewed (changes requested):
Looks pretty good to me. Just a few doc-related suggested changes, and requesting an additional test for the `build` portion.

Shrews commented on ansible_builder/main.py:
```diff
@@ -39,7 +39,19 @@ def __init__(self,
+                 galaxy_ignore_signature_status_code=None):
+        """
+        :param str galaxy_keyring: GPG keyring file used by ansible-galaxy to opportunistically validate collection signatures.
+        :param str galaxy_required_valid_signature_count: Number of sigs (prepend + to disallow no sig( required for ansible-galaxy to accept collections.
```
Still have a ( here that should be a ).

eqrx replied:
oh .... did not notice I copy pasted that in 5 locations, thanks!

Shrews commented on docs/usage.rst:
```diff
@@ -75,6 +75,46 @@ To use a definition file named something other than
+``--galaxy-ignore-signature-status-code``
+*****************************************
+
+With ``--galaxy-keyring ``set it is possible to ignore certain errors that may occur while while verifying collections.
```
```suggestion
With ``--galaxy-keyring`` set it is possible to ignore certain errors that may occur while while verifying collections.
```

Shrews commented on test/integration/test_build.py:
```diff
@@ -187,3 +189,48 @@ def test_build_layer_reuse(cli, runtime, data_dir, ee_tag, tmp_path):
+    assert f"--keyring {constants.default_keyring_name}" in result.stdout
+
+
+#@pytest.mark.test_all_runtimes
```
Did you mean to leave this commented out?

eqrx replied:
yes :< I was unsure about fixing a part of it, working on it now

Shrews reviewed (approved):
LGTM. Not sure if @nitzmahone wants to review or not.

nitzmahone commented on ansible_builder/cli.py:
```diff
@@ -143,6 +143,16 @@ def add_container_options(parser):
+        p.add_argument('--galaxy-keyring',
+                       help='Keyring for collection signature verification during installs from Galaxy. '
+                            'Will be copied into images. Verification is disabled if unset.')
+        p.add_argument('--galaxy-ignore-signature-status-code',
```
This is a multi-value `type='append'` option in the Galaxy CLI, so this one probably needs to be as well- then we'll need to generate one `--ignore-signature-status-code` for each value we got. Could probably also update the description to match Galaxy CLI re: specifying multiple values.

nitzmahone commented on ansible_builder/steps.py:
```diff
@@ -43,14 +43,36 @@ def __init__(self):
+        env = ""
+        install_opts = f"-r {requirements_naming} --collections-path {constants.base_collections_path}"
+
+        if galaxy_ignore_signature_status_code:
```
multi-value support needed (see above)

nitzmahone commented on ansible_builder/steps.py:
```diff
@@ -43,14 +43,36 @@ def __init__(self):
         """
+
+        env = ""
+        install_opts = f"-r {requirements_naming} --collections-path {constants.base_collections_path}"
```
We should probably quote these values in the generated cmdline string (eg, paths with spaces)

nitzmahone commented on ansible_builder/steps.py:
```diff
@@ -43,14 +43,36 @@ def __init__(self):
-                requirements_naming, constants.base_roles_path),
-            "RUN ansible-galaxy collection install $ANSIBLE_GALAXY_CLI_COLLECTION_OPTS -r {0} --collections-path {1}".format(
-                requirements_naming, constants.base_collections_path),
+            f"RUN ansible-galaxy role install -r {requirements_naming} --roles-path {constants.base_roles_path}",
```
also quoted arg values

nitzmahone commented on docs/usage.rst:
```diff
@@ -75,6 +75,46 @@ To use a definition file named something other than
+``--galaxy-ignore-signature-status-code``
+*****************************************
+
+With ``--galaxy-keyring`` set it is possible to ignore certain errors that may occur while while verifying collections.
```
s/while while/while/ ;)

nitzmahone commented on docs/usage.rst:
```diff
@@ -75,6 +75,46 @@ To use a definition file named something other than
+``--galaxy-required-valid-signature-count``
+*******************************************
+
+With ``--galaxy-keyring`` set it is possible specify constrainsts of collection validation. The value is
```
```suggestion
When ``--galaxy-keyring`` is set, the number of required valid collection signatures can be overridden. The value is
```

nitzmahone commented on test/integration/test_build.py:
```diff
@@ -187,3 +189,52 @@ def test_build_layer_reuse(cli, runtime, data_dir, ee_tag, tmp_path):
+    ee_def = data_dir / 'ansible.posix.at' / 'execution-environment.yml'
+
+    result = cli(f'ansible-builder build -c {tmp_path} -f {ee_def} -t {ee_tag} --container-runtime {runtime} -v 3 '
+                 f'--galaxy-keyring {keyring} --galaxy-ignore-signature-status-code 500 '
```
might as well use legit ignore codes in here as well (see above)

nitzmahone commented on test/integration/test_build.py:
```diff
@@ -187,3 +189,52 @@ def test_build_layer_reuse(cli, runtime, data_dir, ee_tag, tmp_path):
+                 f'--galaxy-required-valid-signature-count 3', allow_error=True)
+
+    assert "--galaxy-ignore-signature-status-code 500" in result.stdout
+    assert "--galaxy-required-valid-signature-count 3" in result.stdout
```
Should probably add a multi-value test for `--ignore-signature-status-code` as well

nitzmahone reviewed (changes requested):
Sorry for the slow re-review- a couple little things to adjust around the new ignore code behavior, but looks good otherwise.

nitzmahone commented:
I've also inquired about maybe changing Galaxy CLI's parse action on that arg to `extend` instead of `append`, which would let us generate a slightly less verbose cmdline if we set ours to match.

Shrews committed:
Merge branch 'devel' into signing

Shrews committed:
Update constants.py
